src/exo/extern.py

```diff
@@ -0,0 +1,36 @@
+import math
+
+# --------------------------------------------------------------------------- #
+# --------------------------------------------------------------------------- #
+# Extern superclass
+
+
+class Extern_Typecheck_Error(Exception):
+    def __init__(self, msg):
+        self._builtin_err_msg = str(msg)
+
+    def __str__(self):
+        return self._builtin_err_msg
+
+
+_EErr = Extern_Typecheck_Error
+
+
+class Extern:
+    def __init__(self, name):
+        self._name = name
+
+    def name(self):
+        return self._name
+
+    def globl(self, prim_type):
+        raise NotImplementedError()
+
+    def typecheck(self, args):
```

Comment:
I think it is confusing to ask the user to implement type checking. If all we need to do is check the number of arguments, check that all of them are reals, and I think we want to check that all of them are the same precision. We can check all of these things on our own in the frontend and at codegen. You probably want to check all of these in the frontend. Ast codegen, you can recheck all of them as a sanity check, but I think the last condition is the only one that could be invalidated through rewrites.

Reply:
All the examples here happen to have the same arguments, but that's not always the case. I think it's reasonable to use externs to implement a user-defined casting (e.g., cast by clamping instead of our default C casting). So, I think typecheck should be exposed to users, not to constrain the kind of function they want to write.

Reply:
The problem with this is that it contradicts the fact that you cannot do mixed precision operations. You might want to be able to specify the return type to be something else for the case you are talking about, but I still think the parameters should be of the same type. This is especially the case when you specify `def globl(self, prim_type)` is associated with a single type.

Reply:
I think it's totally fine to be able to mix precisions (not in the externs examples we have right now, but why not?). To define a user-defined casting with externs, the argument type and the return type needs to be different.

Reply:
Here is an example where this becomes problematic:

```
@proc
    def foo(a: f32, b: f64, c: i32, d: i8, e: f64):
      e = select(a, b, c, d)
```

which will generate the following on the code from this PR:
```
int8_t _select_int8_t(int8_t x,int8_t v,int8_t y,int8_t z) {
    if (x < v) return y;
    else return z;
}

// foo(
//     a : f32 @DRAM,
//     b : f64 @DRAM,
//     c : i32 @DRAM,
//     d : i8 @DRAM,
//     e : f64 @DRAM
// )
void foo( void *ctxt, const float* a, const double* b, const int32_t* c, const int8_t* d, double* e ) {
*e = (double)(_select_int8_t((int8_t)*a, (int8_t)*b, (int8_t)*c, (int8_t)*d));
}
```
This would compute the wrong thing.

Previously, the implementation implemented `select` with double as the type of all the arguments and casted any argument into that which I thought made sense because it is a best-effort to have an implementation that always has a high-enough precision. Yet, that implementation could have also broken in some cases.

This is why I think unifying the types of the arguments makes sense because we have always said "Exo requires all data-expressions to have consistent precision, (e.g. multiplying an f32 and i8 is forbidden) but inserts type-casts as necessary just before writing or reducing data values." Now, you can generate those extern implementations correctly since there is exactly one type.

Reply:
So I agree that `select`, should, check the type consistency of arguments, but my argument is that there's no reason to impose that constraint to users. Users should be able to write clamp casting with externs.

Reply:
```

class _Clamp(Extern):
    def __init__(self):
        super().__init__("clamp")

    def typecheck(self, args):
        if len(args) != 1:
            raise _EErr(f"expected 1 argument, got {len(args)}")

        atyp = args[0].type
        if not isinstance(atyp, T.f32):
            raise _EErr(f"expected argument 1 to be f32")
        return atyp

    def globl(self, prim_type):
        assert prim_type == "float"
        s = (
            f"i8 _clamp_(float x) " + "{\n"
            "    ...."
            "}\n"
        )
        return s

    def compile(self, args, prim_type):
        assert prim_type == "float"
        return f"_clamp_({args[0]})"

clamp = _Clamp()
```

Reply:
This example still doesn't require mixed precision arguments (the arguments are all of the same type, the return value 
 is different). In any case, if you really think type-checking should be something that is exposed for the users. I would suggest rewrite this class such that it defines a signature of the arguments and return types e.g. get_signature(): return `(ExoType.i8, (ExoType.f32, ))` and we automatically generate a type check based on this i.e. Tuple(ReturnType, Tuple(Arg0Type, Arg1Type, ...)).